packages/scenario/src/client/traversal.ts

```diff
@@ -0,0 +1,65 @@
+import { UnreachableError } from '@odk-web-forms/common/lib/error/UnreachableError.ts';
+import type { AnyNode, RepeatRangeNode, RootNode } from '@odk-web-forms/xforms-engine';
+import type { Scenario } from '../jr/Scenario.ts';
+
+/**
+ * Collects a flat list of a form instance's nodes, in an order consistent with
+ * the linearly indexed progress tracked by {@link Scenario} (as preserved from
+ * JavaRosa). Note that this function does not filter node types, intentionally
+ * preserving nodes which would be excluded from such progress tracking. This is
+ * so the same collection may be used to retrieve those nodes' values in order
+ * to satisfy {@link Scenario.answerOf}, despite those nodes being omitted in
+ * the final tracked linear progress state (and presumably also excluded from
+ * performing writes in {@link Scenario.answer}).
+ */
+export const collectFlatNodeList = (currentNode: AnyNode): readonly AnyNode[] => {
+	switch (currentNode.nodeType) {
+		case 'root':
+		case 'repeat-instance':
+		case 'group':
+		case 'subtree':
+			return [currentNode, currentNode.currentState.children.map(collectFlatNodeList)].flat(2);
+
+		case 'repeat-range':
+			return [currentNode.currentState.children.map(collectFlatNodeList), currentNode].flat(2);
+
+		case 'select':
+		case 'string':
+			return [currentNode];
+
+		default:
+			throw new UnreachableError(currentNode);
+	}
+};
+
+export const getNodeForReference = (instanceRoot: RootNode, reference: string): AnyNode | null => {
+	const nodes = collectFlatNodeList(instanceRoot);
```

Comment:
🔧 can we cache this?

Reply:
Maybe? It would require shuffling a few things around. What would be the goal?

Reply:
To optimize the code. The function `getNodeForReference` is called a lot during the test execution via `answerOf`, I would have created a map of node by reference; Since these are just tests, it is **not** super important to optimize it. 

While going through the code, this showed up as a minor improvement, so thought of posting it. It is just a good faith suggestion - take it or leave it

Reply:
I took a quick look into whether this would be doable with a quick `createMemo` wrapper. Unfortunately that causes a slew of new failures with `InconsistentChildrenStateError`. This is a known issue affecting a few tests already, apparently triggered by `createMemo` in a Solid client when repeat instances are added/removed. While I certainly consider that a bug, I think expanding it to cause a whole lot of other tests to fail would be less beneficial than keeping it isolated to the three tests currently failing.

That said, and _I know I'll sound like a broken record_... we really should try to base optimization targets on measurements. And since I keep repeating this point, I did measure it. On my machine, nearly every call to this from the root (i.e. not a recursion from there) completes in a **tiny fraction** of a millisecond. The absolute worst case is in `who-va.test.ts`, in which:

- Each call averages 1.21ms.
- There are 272 total calls, for a total of 330.07ms. Not nothing, but not what I'd call an enormous test overhead for a notably heavy smoke test.
- Wrapping with a `createMemo` _slows the calls_ by 40ms (and overall slows the full module's runtime down by at least half a second).

Apart from that module, the total time on my machine _for all calls from all other test modules_ is 21.64ms.

I'm sensitive to slow test runtime (like a lot). And to the fact that the `who-va.test.ts` suite in particular is slow, to the point of being flaky. But I think we'll probably get more value out of improving the **performance of the engine** than optimizing this aspect of test orchestration.